src/tir/schedule/primitive/blockize_tensorize.cc

```diff
@@ -0,0 +1,638 @@
+  /*! \brief Loop variables of the outer loops. */
+  Array<Var> outer_loop_vars;
+  /*! \brief Loop variables of the inner loops. */
+  Array<Var> inner_loop_vars;
+  /*! \brief Domain of the loop variables. */
+  Map<Var, Range> loop_var_domain;
+};
+
+/*!
+ * \brief Check the bindings of the block iters can be divided by a subspace collected by the
+ * collector.
+ * \param mod The current IR module.
+ * \param block_realize The block realize to be checked.
+ * \param collector The collector which has collected the loops of the block.
+ * \param analyzer The arithmetic analyzer.
+ * \return The result of the subspace division.
+ * \throws ScheduleError If the bindings are not divisible by the subspace.
+ */
+Array<Array<arith::IterMark>> CheckSubspaceDivisible(const IRModule& mod,
+                                                     const BlockRealize& block_realize,
+                                                     const LoopSubspaceCollector& collector,
+                                                     arith::Analyzer* analyzer) {
+  const Block& block = block_realize->block;
+  DiagnosticContext diag_ctx(DiagnosticContext::Default(mod));
+
+  Array<Array<arith::IterMark>> division =
+      arith::SubspaceDivide(block_realize->iter_values, collector.loop_var_domain,
+                            collector.inner_loop_vars, block_realize->predicate,
+                            /*require_bijective=*/false, analyzer, diag_ctx);
+
+  if (division.empty()) {
+    // If we can't do perfect subspace division, check if it is a trivial case of subspace division.
+    // In this case, we can still blockize.
+    division = TrivialSubspaceDivision(block->iter_vars, block_realize->iter_values,
+                                       collector.outer_loop_vars, collector.inner_loop_vars,
+                                       block_realize->predicate);
+  }
+  if (division.empty()) {
+    throw SubspaceNotDivisibleError(mod, GetRef<For>(collector.inner_loops.back()), block);
+  }
+  return division;
+}
+
+/*!
+ * \brief The binding extractor to compute the bindings of the outer and the inner blocks after
+ * blockize.
+ */
+class BlockizedBindingExtractor {
+ public:
+  /*!
+   * \brief Extract bindings for blockize.
+   * \param iter_vars The iter vars of the original inner block.
+   * \param division The result of the subspace division.
+   */
+  void ExtractBindings(const Array<IterVar>& iter_vars,
+                       const Array<Array<arith::IterMark>>& division) {
+    ICHECK(iter_vars.size() + 1 == division.size());
+    for (size_t i = 0; i < iter_vars.size(); ++i) {
+      const IterVar& iter_var = iter_vars[i];
+      const arith::IterMapExprNode* outer_binding =
+          division[i][0]->source.as<arith::IterMapExprNode>();
+      const arith::IterMapExprNode* inner_binding =
+          division[i][1]->source.as<arith::IterMapExprNode>();
+      ICHECK(outer_binding);
+      ICHECK(inner_binding);
```

Comment:
Use `IterSumExprNode` instead to be more accurate. `TVM_TYPE_AS` could help with type checking BTW.

```suggestion
      const auto* outer_binding = TVM_TYPE_AS(outer_binding, division[i][0]->source, arith::IterSumExprNode);
      const auto* inner_binding = TVM_TYPE_AS(inner_binding, division[i][1]->source, arith::IterSumExprNode);
```